bionic/code_references.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2018-2020 Streamlit Inc.
+# Author: Dominik Moritz
+
+"""
+Contains the logic to find all the referenced objects for a code object.
+"""
+
+import attr
+import dis
+import importlib
+import inspect
+import warnings
+
+from .utils.misc import oneline
+
+
+@attr.s(frozen=True)
+class CodeContext:
+    globals = attr.ib()
+    cells = attr.ib()
+    varnames = attr.ib()
+
+
+def get_code_context(func) -> CodeContext:
+    code = func.__code__
+
+    # Mapping from variable name to the value if we can resolve it.
+    # Otherwise map to the name.
+    cells = {}
+
+    for var in code.co_cellvars:
+        cells[var] = var  # Instead of value, we use the name.
+
+    if code.co_freevars:
+        assert len(code.co_freevars) == len(func.__closure__)
+        cells.update(
+            zip(code.co_freevars, map(lambda c: c.cell_contents, func.__closure__))
+        )
+
+    varnames = {}
+    if inspect.ismethod(func):
+        varnames = {"self": func.__self__}
+
+    return CodeContext(globals=func.__globals__, cells=cells, varnames=varnames)
+
+
+def get_referenced_objects(code, context):
```

Comment:
1. What is the definition of "major"? And why?
2. If we don't catch every kind of reference, why not?
3. How do we know that it will catch every major reference? I'd like to be more confident than "I tried as many tests as I could think of and they passed". We should be able to read the code and be convinced that it is correct. What can you tell the reader that would convince them?

Reply:
I added docstring to this method. But I think it's really hard to justify `3.` without explaining how python bytecode instructions work.

Reply:
I added some more details in the function docstring.